### ISSUE TITLE
You Only Look Once: Unified, Real-Time Object Detection (https://arxiv.org/pdf/1506.02640.pdf)

### ISSUE THREAD

**Issue by vageeshSaxena** [vageeshSaxena/Deep-Learning#1]
**Objective**: To make an end-to-end network that can frame object detection as a regression problem to spatially separated bounding boxes and associated class probabilities.

**State-of-art-models**:
1) To detect an object, these systems take a **classifier** for that object and evaluate it at **various locations and scales in a test image**.
2) Systems like **deformable parts models (DPM)** use a **sliding window approach** where the classifier is run at evenly spaced locations over the entire image. 
I) DPM uses a disjoint pipeline to extract static features, classify regions, predict bounding boxes for high scoring regions, etc. 
3) **R-CNN's** use **region proposal method**s to first **generate potential bounding boxes** in an image and then **run a classifier** on these proposed boxes. After classification, post-processing is used to refine the bounding boxes, eliminate duplicate detections, and re-score the boxes based on other objects in the scene. 
I) Selective search generates potential bounding boxes, a convolutional network extracts features, 
an SVM scores the boxes, a linear model adjusts the bounding boxes, and non-max suppression 
eliminates duplicate detections.
II) Each stage of this complex pipeline must be precisely tuned independently and the resulting system is very slow, taking more than 40 seconds per image at test time.
4) **Fast R-CNN, a top detection method, mistakes background patches in an image for objects because it can’t see the larger context.**
I) Fast and Faster R-CNN focus on speeding up the R-CNN framework by sharing computation and using neural networks to propose regions instead of Selective Search.
II) While they offer speed and accuracy improvements over R-CNN, both still fall short of real-time performance.
5) **Deep MultiBox** uses a **convolutional neural network to predict regions of interest** instead of using Selective Search.
I) MultiBox can also perform single object detection by replacing the confidence prediction with a single class prediction.
II) **MultiBox cannot perform general object detection** and is still just a piece in a larger detection pipeline, requiring further image patch classification.
6) **OverFeat** uses a **convolutional neural network** to perform **localization and adapt that localizer to perform detection.**
I) OverFeat efficiently performs sliding window detection but it is still a disjoint system.
II)  OverFeat optimizes for localization, not detection performance.
III) Like DPM, the **localizer only sees local information** when making a prediction. OverFeat cannot reason about global context and thus requires significant post-processing to produce coherent detections.
7) **MultiGrasp uses bounding box prediction** for **regression** to grasps.
I) Grasp detection is a much simpler task than object detection. MultiGrasp only needs to predict a single graspable region for an image containing one object. It doesn’t have to estimate the size, location,
or boundaries of the object or predict it’s class, only find a region suitable for grasping. 
8) These complex pipelines are **slow and hard** to optimize because each individual component must be trained separately.

**Network Design**:
1) Inspired by the GoogLeNet model for image classification, the network has **24 convolutional layers followed by 2 fully connected layers**.
2) Instead of the inception modules used by GoogLeNet, **1 × 1 reduction layers** is used followed by **3 × 3 convolutional layers**.
3) The paper also talks about Fast YOLO, which has the same network architecture but less deeper. Fast YOLO uses a neural network with fewer convolutional layers (9 instead of 24) and fewer filters in those layers. Other than the size of the network, all training and testing parameters are the same between YOLO and Fast YOLO.

**Architecture properties**: 
1) The architecture is extremely fast.
2) **Baseline** processes **45 frames/sec** for real-time images with no batch processing.
3) A small version of the network, **FAST YOLO**, processes **155 frames/second**.
4) **YOLO** makes **more localization errors** but is less likely to predict **false positives** on the background.
5) **YOLO** achieves **more than twice** the **mean average precision** of other real-time systems. 
6) **YOLO** makes less than **half** the number of **background errors** compared to **Fast R-CNN**.
7) YOLO learns generalizable representations of objects.
8) **YOLO** outperforms **top detection methods** like **DPM and R-CNN** by a wide margin but it **lags** behind state-of-the-art detection systems in **accuracy**.

**How does YOLO works?**:
1)  A **single convolutional network** simultaneously predicts multiple bounding boxes and class probabilities for those boxes.
2) Unlike sliding window and region proposal-based techniques, YOLO uses features from the **entire image** to predict bounding boxes during training and test time so it implicitly **encodes contextual information about classes** as well as their appearance.
3) The initial **convolutional layers** of the network **extract features from the image** while the **fully connected layers** predict the **output probabilities and coordinates.**
4) The network divides the **input image into an S × S grid**.  **If the centre of an object falls into a grid cell, that grid cell is responsible for detecting that object.**
5) **Each grid cell predicts B bounding boxes and confidence scores for those boxes**. These confidence scores reflect how confident the model is that the box contains an object and also how accurate it thinks the box is that it predicts. If no object exists in that cell, the confidence scores should be zero. Otherwise, we want the confidence score to equal the intersection over union (IOU) between the predicted box and the ground truth.
6) **Each bounding box consists of 5 predictions: x, y, w, h, and confidence**. The (x, y) coordinates represent the center of the box relative to the bounds of the grid cell. The width and height are predicted relative to the whole image. Finally, the confidence prediction represents the IOU between the predicted box and any ground truth box.

**Training** - 
1) Pre-training of the convolutional layers are done **on the ImageNet classification task** at half the resolution (224 × 224 input image) and then the resolution are doubled for detection.
2) For pretraining, the **first 20 convolutional layers** are used followed by an **average-pooling layer and a fully connected layer.** 
3) **Accuracy :** 88% on the ImageNet 2012 validation set.
4) **Framework used :**  Darknet.
5) 4 convolutional layers and 2 fully connected layers are then added with randomly initialized weights.
6) The final layer predicts both class probabilities and bounding box coordinates. The bounding box width and height has been normalized by the image width and height so that they fall between 0 and 1. 
7) **Activation function : ReLu** for all the other layers and **linear acitvation function** for the final one.
8) **Sum-squared error** is used to optimization. However it does not perfectly align with the goal of
maximizing average precision. It weights localization error equally with classification error which may not be ideal. Also, in every image many grid cells do not contain any object. This pushes the “confidence” scores of those cells towards zero, often overpowering the gradient from cells that do contain objects. This can lead to model instability, causing training to diverge early on. To remedy this, loss has been increased from bounding box coordinate predictions and decrease the loss from confidence predictions for boxes that don’t contain objects.
9) Sum-squared error also equally weights errors in large boxes and small boxes. The error metric should reflect that small deviations in large boxes matter less than in small boxes. To partially address this, the implementation predict the square root of the bounding box width and height instead of the width and height directly.
10) YOLO predicts multiple bounding boxes per grid cell. At training time, only one bounding box predictor is responsible for each object. One predictor to be “responsible” for predicting an object based on which prediction has been assigned the highest current IOU with the ground truth. This leads to specialization between the bounding box predictors. Each predictor gets better at predicting certain
sizes, aspect ratios, or classes of object, improving overall recall.
11)  The loss function only penalizes classification error if an object is present in that grid cell (hence the conditional class probability discussed earlier). It also only penalizes bounding box coordinate error if that predictor is“responsible” for the ground truth box  (i.e. has the highest IOU of any predictor in that grid cell).
12) **Throughout training the implemenation uses a batch size of 64, a momentum of 0.9 and a decay of 0.0005. The network is trained for about 135 epochs on the training and validation data sets from PASCAL VOC 2007 and 2012.**
13) For the first epochs, the **learning rate is** increased from 10^-3 to **10^−2**. For high learning rates, the model often diverges due to unstable gradients. Training si done with 10^−2 for 75 epochs, then 10^−3
for 30 epochs, and finally **10^−4** for 30 epochs.
14) **To avoid overfitting, Dropouts and extensive data augmentation are used.** A dropout layer with rate = .5 after the first connected layer prevents co-adaptation between layers. For data augmentation, random scaling and translations of up to 20% of the original image size. The implementation also randomly adjust the exposure and saturation of the image by up to a factor of 1.5 in the HSV color space.

**Evaluation** :
1) **Test dataset :** PASCAL VOC.
2) Predicting detections for a test image only requires one network evaluation. **The network predicts 98 bounding boxes per image and class probabilities for each box.**  
3) The grid design enforces spatial diversity in the bounding box predictions. Often it is clear which grid cell an object falls in to and the network only predicts one box for each object. However, some large objects or objects near the border of multiple cells can be well localized by multiple
cells. Non-maximal suppression can be used to fix these multiple detections. While not critical to performance as it is for R-CNN or DPM, non-maximal suppression adds 2- 3% in mAP.

**Limitations** :
1) YOLO imposes strong spatial constraints on bounding box predictions since each grid cell only predicts two boxes and can only have one class. This spatial constraint limits the number of nearby objects that our model can predict.**The model struggles with small objects that appear in
groups, such as flocks of birds.**
2) Since the model learns to predict bounding boxes from data, **it struggles to generalize to objects in new or unusual aspect ratios or configurations.**
3) The training on the loss function that approximates detection performance treats errors the same in small bounding boxes versus large bounding boxes. **A small error in a large box is generally benign but a small error in a small box has a much greater effect on IOU.**The main source of error is **incorrect localizations.**

**Combining Fast R-CNN and YOLO**
1) YOLO makes far fewer background mistakes than Fast R-CNN. By using **YOLO** to **eliminate** **background detections** from Fast R-CNN, one can get a significant boost in performance. **For every bounding box that R-CNN predicts, we can check to see if YOLO predicts a similar box. If it does, we give that prediction a boost based on the probability predicted by YOLO and the overlap between the two boxes.**
2) **The boost from YOLO** is not simply a byproduct of model ensembling since there is little benefit from combining different versions of Fast R-CNN. Rather, it i**s precisely because YOLO makes different kinds of mistakes at test time that it is so effective at boosting Fast R-CNN’s performance.**
3) Unfortunately, this combination **doesn’t benefit from the speed of YOLO** since each model is run seperately and then the results are combined. However, since YOLO is so fast it doesn’t add any significant computational time compared to Fast R-CNN.

**Check the tables 3 and 5 in the paper for results.** 
Spoiler - Surprise surpise, YoLo+Fast R-CNN performs the best.
